gas.md

```diff
@@ -513,6 +518,89 @@ Definitions
+    * `time z1 = time t + 2 * add (2 * n + 2)` and `size z1 = 2 * n + 2`
+    * Since the shifting is done by limbs, we can aggregate the cost for shifting
+      into that for addition, and have:
+
+      ```hs
+      mul (2*n) = time (x * y) = time z0 + time z1 + time z2 + 2 * add (4 * n) + const0
+                 = 2 * mul n + mul (n + 1) + 2 * add n + 2 * add (2 * n + 2) + 2 * add (4 * n) + const0
+      size x * y = 4 * n
+      ```
+
+  * Using that add is linear, we can assume constants `cAdd1` and `cAdd0` s.t.
+    `add x = cAdd1 * x + cAdd0`, we can rewrite `mul (2*n)` to:
+
+    `mul (2*n) = 2 * mul n + mul (n + 1) + 2 * (cAdd1 * n + cAdd0) + 2 * (cAdd1 * (2 * n + 2) + cAdd0) + 2 * (cAdd1 * 4 * n + cAdd0) + const0`, which simplifies to
+    ```hs
+    mul n = 2 * mul (nd2 1)  + mul (nd2 1 + 1) + n * cAdd1 * 7 + const1
+       where nd2 k = ceiling (n / 2 ^ k)
+             const1 = const0 +  cAdd0 * 6 + cAdd1 * 4
+    ```
+
+    Now, to simplify computation, we approximate `mul (nd2 1)` with `mul (nd2 1 + 1)`,
+    and `ceiling ((nd2 1 + 1) / 2)` with `nd2 2`
+    ```hs
+    mul n = 3 * mul (nd2 1 + 1) + 7 * cAdd1 * n + const1
+          = 3 * (3 * mul (nd2 2 + 1) + 7 * cAdd1 * (nd2 1 + 1) + const1) + 7 * cAdd1 * n + const1
+    ```
+    Let `const2 = 7 * cAdd1 + const1`. Then we can replace `const1` by `const2 - 7 * cAdd1`
+    ```hs
+    mul n = - 7 * cAdd1 + 7 * cAdd1 * n + const2
+                       + (7 * cAdd1 * nd2 1 + const2) * 3 ^ 1
+                       + (7 * cAdd1 * nd2 2 + const2) * 3 ^ 2 +
+                       .................
+                       + (7 * cAdd1 * nd2 k + const2) * 3 ^ k
+      where k = log n / log 2
+    ```
+    By computing the sums, we obtain
+    ```hs
+    mul n = -7 * cAdd1 + 7 * cAdd1 * n * ((3/2)^(k+1) -1)(3/2 - 1) + const2*(3 ^ (k+1) -1) / 2
+    ```
+    Considering that `2^k = n` and `3 ^ k = n ^ constKara`, we reduce to
+    ```hs
+    mul n = 14 * cAdd1 * n * (3/2 * n ^ constKara / n - 1) + const2/2 * (3 * n^constKara - 1) - 7 * cAdd1
+          = 21 * n ^ constKara + 3/2 * const2 * n^constKara - 14 * cAdd1 * n - 7 * cAdd1 - const2/2
+          = mMul2 * n ^ constKara + mMul1 * n + mMul0
+      where mMul2 = 21 + 3 * const2 / 2 = 21 + 3 / 2 * (11 * cAdd1 + 6 * cAdd0 + const0)
+            mMul1 = -14 * cadd1
+            mMul0 = -7*cAdd1 - const2 / 2 = -1/2 * (25 * cAdd1 + 6 * cAdd0 + const0)
+            const2 = 11 * cAdd1 + 6 * cAdd0 + const0
+    ```
+* Division cost: `a divmod m` when size of `a` is twice the size of `m`, say `n` division
+  can be done in a divide and conquer manner, with a complexity of `mul n * log n`
+  ```hs
+  divModCost n = mul n * (log n / log 2)
+  ```
+* Multiplication modulo cost. Assuming size of arguments is equal to size of modulo, which is `n`
+  ```hs
+  mulModCost n = mul n + divCost (2 * n) n = mul n * (1 + log n)
+  ```
+  When doing multiple multiplications modulo `m` of size `n` for really large `n`,
```

Comment:
I was just thinking about something, I hope I'm not too tired yet and what I'm writing below makes sense.

You could do Karatsuba for multiplications modulo-m:
Let's say that you want to compute x*y(mod m), `x = b^n * x1 + x0,  y = b^n * y1 + y0`. Let us assume that m > b^2, otherwise the problem is much easier.

Then, you can try to do that by computing p = (x1 + x2)*(y1 + y2) (mod m), q = x1*y1 (mod m), r = x2*y2 (mod m), s = b^2n (mod m), t = b^n(mod m) and computing x*y by doing q*s (mod m) + (p-q-r) * t + r.

One has to apply this algorithm until x, y < sqrt(m) (or something similar), then fall back to the normal Karatsuba.

While this seems to have more multiplications than Karatsuba (i.e. q*s (mod m)), note that these extra multiplications have limited terms (i.e. <= m). The number of these extra multiplications is O(log(|x|-|m|)) where |k| is the length of k in base b.

This may provide reasonable complexity on both cases.